fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/api/recipe/v1/ingredient/DefaultCustomIngredients.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.recipe.v1.ingredient;
+
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.nbt.NbtHelper;
+import net.minecraft.recipe.Ingredient;
+
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.AndIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.DifferenceIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.NbtIngredient;
+import net.fabricmc.fabric.impl.recipe.ingredient.builtin.OrIngredient;
+
+/**
+ * Factory methods for the custom ingredients directly provided by Fabric API.
+ */
+public final class DefaultCustomIngredients {
+	/**
+	 * Creates an ingredient that matches when its sub-ingredients all match.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *     "fabric:type": "fabric:and",
+	 *     "ingredients": [
+	 *         // sub-ingredient 1,
+	 *         // sub-ingredient 2,
+	 *         // etc...
+	 *     ]
+	 * }
+	 * }</pre>
+	 *
+	 * @throws IllegalArgumentException if the array is empty
+	 */
+	public static Ingredient and(Ingredient... ingredients) {
+		for (Ingredient ing : ingredients) Objects.requireNonNull(ing, "Ingredient cannot be null");
+
+		return new AndIngredient(ingredients).toVanilla();
+	}
+
+	/**
+	 * Creates an ingredient that matches when any of its sub-ingredients matches.
+	 *
+	 * <p>The JSON format is as follows:
+	 * <pre>{@code
+	 * {
+	 *     "fabric:type": "fabric:or",
+	 *     "ingredients": [
+	 *         // sub-ingredient 1,
+	 *         // sub-ingredient 2,
+	 *         // etc...
+	 *     ]
+	 * }
+	 * }</pre>
+	 *
+	 * @throws IllegalArgumentException if the array is empty
+	 */
+	public static Ingredient or(Ingredient... ingredients) {
```

Comment:
Nit: Maybe call this `any`?

Reply:
Could also be `union`, not sure what's best.

Reply:
I'd argue for `any`/`all` since they're simpler. `union` and `intersection` seem a bit out of place as more set-theoretical terms.